perp/man/perphup.8
.\" perphup.8
.\" wcm, 2009.12.01 - 2011.03.18
.\" ===
.TH perphup 8 "March 2011" "perp-2.04" "persistent process supervision"
.SH NAME
perphup \- trigger a
.BR perpd (8)
rescan
.SH SYNOPSIS
.B perphup [\-hV] [\-q] [\-t] [
.I basedir
.B ]
.SH DESCRIPTION
.B perphup
is used to trigger
.BR perpd (8)
to rescan its directory of service definitions.
.PP
.B perphup
looks for an instance of
.BR perpd (8)
running on the argument given by
.IR basedir .
Otherwise,
.B perphup
looks for a value in the environmental variable PERP_BASE.
If neither of these are defined or are empty,
.B perphup
uses a compiled-in default, normally
.IR /etc/perp .
.PP
When
.B perphup
finds the 
.BR perpd (8)
process, it sends it a SIGHUP signal.
.SH OPTIONS
.TP
.B \-h
Help.
Print a brief usage message to stderr and exit.
.TP
.B \-q
Quiet.
Normally after successfully finding and triggering a
.BR perpd (8)
process,
.B perphup
prints a brief message to stderr.
The
.B \-q
option may be used to suppress this message.
.TP
.B \-t
Term.
Normally
.B perphup
sends a SIGHUP signal to the
.BR perpd (8)
process to trigger a rescan.
The
.B \-t
option is used to send a SIGTERM instead,
and triggers the
.BR perpd (8)
shutdown sequence.
.TP
.B \-V
Version.
Print the version number to stderr and exit.
.SH EXIT STATUS
.B perphup
exits with the following values:
.TP
0
Success.
An active
.BR perpd (8)
process was found and triggered.
.TP
100
Usage error.
For unknown options or other command-line errors.
Prints a brief diagnostic message to stderr on exit.
.TP
111
System error.
Unexpected failures during system calls, privilege and/or resource problems,
or configuration problems in the
.BR perpd (8)
directory.
Prints a brief diagnostic message to stderr on exit.
.SH AUTHOR
Wayne Marshall, http://b0llix.net/perp/
.SH SEE ALSO
.nh
.BR perp_intro (8),
.BR perpboot (8),
.BR perpctl (8),
.BR perpd (8),
.BR perpetrate (5),
.BR perpls (8),
.BR perpok (8),
.BR perpstat (8),
.BR sissylog (8),
.BR tinylog (8)
.\" EOF perphup.8


CHANGES
# CHANGES
# perp release history and changelog
# wcm, 2010.02.17 - 2011.03.22
# ===

perp-2.04 (2011.03.22):
 * perphup:
   - added option -t, trigger perpd shutdown
 * perpboot:
   - waitpid() for deux[1] on SIGTERM
 * perp-setup:
   - added arg for ctldir
   - configure symlink /etc/perp/.control --> ctldir
   - configure rc.perp bootscript for runchoom(8)
 * runchoom:
   - new, utility for linux "oom killer" abatement

perp-2.03 (2011.03.11):
 * perp-setup:
   - in rc.perp boot script for perpd(8), setup default runlimit(8)
     configuration to hard limits for RLIMIT_NPROC, RLIMIT_NOFILE

perp-2.02 (2011.02.22):
 * perpd:
   - eliminated some spurious/redundant client connection logging

perp-2.01 (2011.02.11):
 * perp-setup:
   - runlimit added to .boot/rc.perp script
 * perpd:
   - added option -g, enable group permission on control socket
   - additional comments in LIMITS section of perpd(8) manual
 * runlimit:
   - added option -E, take resource limits from environment
   - added option -F, take resource limits from file
 * lasagna:
   - outvec/mk_outvec.c: portability fix for dietlibc

perp-2.00 (2011.02.02):
 * perpd:
   - coalesce supervisor (formerly perpetrate(8)) into scanner/controller
   - runs as a single process
   - all ipc with client applications via single domain socket 
   - ipc protocol version 2 (see PROTO_V2.txt)
 * perpetrate:
   - eliminate perpetrate(8) executable (subsumed by perpd(8))
   - perpetrate(5) manual page retained for runscript conventions
 * perpctl:
   - add commands [A | X], activate/deactivate service
 * libasagna:
   - nuscan.h, improve numeric scanning interface (supercedes nscan.h)
   - other iterative improvements incorporated from projects hdb and ezcdb
 
perp-0.00 (2010.01.12):
 - maiden public release

### EOF


perp/perphup.c
/* perphup.c
** perp: persistent process supervision
** perp 2.0: single process scanner/supervisor/controller
** perphup: trigger rescan in perpd
** wcm, 2009.11.11 - 2011.01.24
** ===
*/

/* libc: */
#include <stdlib.h>

/* unix: */
#include <unistd.h>
#include <errno.h>
#include <signal.h>

/* lasagna: */
#include "cstr.h"
#include "nextopt.h"
#include "nfmt.h"
#include "outvec.h"
#include "pidlock.h"
#include "sysstr.h"

/* perp: */
#include "perp_common.h"
#include "perp_stderr.h"


/* logging variables in scope: */
static const char *progname = NULL;
static const char  prog_usage[] = "[-hV] [-q] [basedir]";

/* option variables in scope: */
/* "quiet", !opq_q == verbose, default is verbose: */
static int opt_q = 0;


int
main(int argc, char *argv[])
{
  nextopt_t    nopt = nextopt_INIT(argc, argv, ":hVq");
  char         opt;
  const char  *basedir = NULL;
  char         pathbuf[256];
  size_t       n;
  pid_t        lockpid;
  char         nbuf[NFMT_SIZE];

  progname = nextopt_progname(&nopt);
  while((opt = nextopt(&nopt))){
      char optc[2] = {nopt.opt_got, '\0'}; 
      switch(opt){
      case 'h': usage(); die(0); break;
      case 'V': version(); die(0); break;
      case 'q': ++opt_q; break;
      case ':':
          fatal_usage("missing argument for option -", optc);
          break;
      case '?':
          if(nopt.opt_got != '?'){
              fatal_usage("invalid option -", optc);
          }
          /* else fallthrough: */
      default : die_usage(); break;
      }
  }

  argc -= nopt.arg_ndx;
  argv += nopt.arg_ndx;

  basedir = *argv;

  if(!basedir)
      basedir = getenv("PERP_BASE");
  if(!basedir || (basedir[0] == '\0'))
      basedir = PERP_BASE_DEFAULT;

  if(chdir(basedir) != 0){
      fatal_syserr("failure chdir() to ", basedir);
  }

  /* get pid from active perpd pidlock: */
  n = cstr_vlen(basedir, "/", PERP_CONTROL, "/", PERPD_PIDLOCK);
  if(!(n < sizeof pathbuf)){
      errno = ENAMETOOLONG;
      fatal_syserr("failure locating perpd lockfile ",
                   basedir, "/", PERP_CONTROL, "/", PERPD_SOCKET);
  }
  cstr_vcopy(pathbuf, basedir, "/", PERP_CONTROL, "/", PERPD_PIDLOCK);

  lockpid = pidlock_check(pathbuf);

  if(lockpid == -1){
      fatal_syserr("failure checking lock on ", pathbuf);
  }

  if(lockpid == 0){
      fatal(111, "perpd not running on ", basedir, ": no lock active on ", pathbuf);
  }
 
  if(kill(lockpid, SIGHUP) == -1){
      fatal_syserr("failure kill() on SIGHUP to perpd pid ",
                   nfmt_uint32(nbuf, (uint32_t)lockpid),
                   "running on ", basedir);
  }

  /* success: */
  if(!opt_q){
      eputs(progname, ": perpd rescan triggered on ", basedir);
  }

  die(0);
}


/* eof: perphup.c */
